builder/bundleset_test.go

```diff
@@ -0,0 +1,195 @@
+	tests := []struct {
+		Contents         []byte
+		ExpectedIncludes []string
+		ExpectedPackages []string
+		ExpectedError    bool
+	}{
+		{
+			Contents: []byte(`# Simple fake bundle
+include(a)
+include(b)
+pkg1     # Comment
+pkg2
+`),
+			ExpectedIncludes: []string{"a", "b"},
+			ExpectedPackages: []string{"pkg1", "pkg2"},
+		},
+
+		// Error cases.
+		{Contents: []byte(`include(`), ExpectedError: true},
+		{Contents: []byte(`()`), ExpectedError: true},
+		{Contents: []byte(`incl` + `ud(`), ExpectedError: true}, // Hide the misspelling from linter.
+		{Contents: []byte(`include())`), ExpectedError: true},
+		{Contents: []byte(`include(abc))`), ExpectedError: true},
+	}
+
+	for _, tt := range tests {
+		includes, packages, err := parseBundle(tt.Contents)
+		if err != nil {
+			if tt.ExpectedError {
+				continue
+			}
+			t.Errorf("unexpected error parsing bundle: %s\nCONTENTS:\n%s", err, tt.Contents)
+			continue
+		}
+		if tt.ExpectedError {
+			t.Errorf("unexpected success when parsing bundle\nCONTENTS:\n%s\nPARSED INCLUDES: %s\nPARSED PACKAGES:\n%s", tt.Contents, includes, packages)
+			continue
+		}
+		sameIncludes := reflect.DeepEqual(includes, tt.ExpectedIncludes)
+		samePackages := reflect.DeepEqual(packages, tt.ExpectedPackages)
+		if !sameIncludes || !samePackages {
+			msg := fmt.Sprintf("wrong result when parsing bundle\nCONTENTS:\n%s", tt.Contents)
+			if !sameIncludes {
+				msg += fmt.Sprintf("\nPARSED INCLUDES (%d): %s\nEXPECTED INCLUDES (%d): %s", len(includes), includes, len(tt.ExpectedIncludes), tt.ExpectedIncludes)
+			}
+			if !samePackages {
+				msg += fmt.Sprintf("\nPARSED PACKAGES (%d):\n%s\nEXPECTED PACKAGES (%d):\n%s", len(packages), packages, len(tt.ExpectedPackages), tt.ExpectedPackages)
+			}
+			t.Errorf(msg)
+			continue
+		}
+	}
+}
+
+func TestParseBundleSet(t *testing.T) {
+	type FilesMap map[string]string
+	type CountsMap map[string]int
+
+	// Repurpose empty CountsMap as error to make the test entries less verbose, since we can avoid
+	// putting the attribute names in the literal map.
+	Error := CountsMap{}
+
+	// Replace spaces with new lines.
+	Lines := func(s string) string {
+		return strings.Replace(s, " ", "\n", -1)
+	}
+
+	tests := []struct {
+		Name  string
+		Files FilesMap
+		// TODO: Replace this with the actual map.
+		ExpectedAllPackageCounts CountsMap
+	}{
+		{
+			"simple include",
+			FilesMap{
+				"a": Lines("A1 A2"),
+				"b": "include(a)",
+			},
+			CountsMap{
+				"a": 2,
+				"b": 2,
+			},
+		},
+		{
+			"redundant includes",
+			FilesMap{
+				"a": Lines("A1 A2 A3 A4"),
+				"b": Lines("include(a) B1 B2 B3 B4"),
+				"c": Lines("include(b) C1 C2 C3 C4"),
+				"d": Lines("include(a) include(b) include(c) A1 B1 C1 D1"),
+			},
+			CountsMap{
+				"a": 4,
+				"b": 8,
+				"c": 12,
+				"d": 13,
+			},
+		},
+
+		{
+			"all packages don't have duplicates",
+			FilesMap{
+				"a": Lines("A A A A"),
+				"b": Lines("include(a) A A A A"),
+				"c": Lines("include(b) A A A A include(a)"),
```

Comment:
are we calling this valid? I guess I don't see why not...
```
include(b)
A
A
A
A
include(a)
```

Reply:
Includes could be throw in anywhere, we just like to keep them organized.

Reply:
We could make it invalid later, but at the moment, we play by m4 rules (the program that parses the bundles).